src/WeakValueDict.jl

```diff
@@ -0,0 +1,222 @@
+Base.copy(d::WeakValueDict) = WeakValueDict(d)
+
+WeakValueDict(ps::Pair{K,V}...)           where {K,V} = WeakValueDict{K,V}(ps)
+WeakValueDict(ps::Pair{K}...)             where {K}   = WeakValueDict{K,Any}(ps)
+WeakValueDict(ps::(Pair{K,V} where K)...) where {V}   = WeakValueDict{Any,V}(ps)
+WeakValueDict(ps::Pair...)                            = WeakValueDict{Any,Any}(ps)
+
+function WeakValueDict(kv)
+    try
+        Base.dict_with_eltype((K, V) -> WeakValueDict{K, V}, kv, eltype(kv))
+    catch
+        if !Base.isiterable(typeof(kv)) || !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("WeakValueDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow()
+        end
+    end
+end
+
+function _cleanup_locked(h::WeakValueDict)
+    if h.dirty
+        h.dirty = false
+        idx = Base.skip_deleted_floor!(h.ht)
+        while idx != 0
+            if h.ht.vals[idx].value === nothing
+                Base._delete!(h.ht, idx)
+            end
+            idx = Base.skip_deleted(h.ht, idx + 1)
+        end
+    end
+    return h
+end
+
+Base.sizehint!(d::WeakValueDict, newsz) = sizehint!(d.ht, newsz)
+empty(d::WeakValueDict, ::Type{K}, ::Type{V}) where {K, V} = WeakValueDict{K, V}()
+
+IteratorSize(::Type{<:WeakValueDict}) = SizeUnknown()
+
+islocked(wkh::WeakValueDict) = Base.islocked(wkh.lock)
+lock(f, wkh::WeakValueDict) = Base.lock(f, wkh.lock)
+trylock(f, wkh::WeakValueDict) = Base.trylock(f, wkh.lock)
+
+function Base.setindex!(wkh::WeakValueDict{K}, v, key) where K
+    !isa(key, K) && throw(ArgumentError("$(Base.limitrepr(key)) is not a valid key for type $K"))
+    # 'nothing' is not valid both because 'finalizer' will reject it,
+    # and because we therefore use it as a sentinel value
+    v === nothing && throw(ArgumentError("`nothing` is not a valid WeakValueDict key"))
+    lock(wkh) do
+        _cleanup_locked(wkh)
+        k = getkey(wkh.ht, key, nothing)
+        # The object gets the finalizer no matter if the key is already there or not.
+        finalizer(wkh.finalizer, v)
+        if k === nothing
+            k = key
+        end
+        wkh.ht[k] = WeakRef(v)
+    end
+    return wkh
+end
+function get!(wkh::WeakValueDict{K}, key, default) where {K}
+    v = lock(wkh) do
+        if key !== nothing && haskey(wkh.ht, key)
+            wkh.ht[key].value
+        else
+            wkh[key] = WeakRef(default)
+            default
+        end
+    end
+    return v
+end
+function Base.get!(default::Base.Callable, wkh::WeakValueDict{K}, key) where {K}
+    v = lock(wkh) do
+        _cleanup_locked(wkh)
+        if key !== nothing && haskey(wkh.ht, key)
+            wkh.ht[key].value
+        else
+            wkh[key] = default()
+        end
+    end
+    return v
+end
+
+function Base.getkey(wkh::WeakValueDict{K}, kk, default) where K
+    k = lock(wkh) do
+        k = getkey(wkh.ht, kk, nothing)
+        k === nothing && return nothing
+        return k
```

Comment:
Huh, am I missing something, isn't this the same as this?
```suggestion
        getkey(wkh.ht, kk, nothing)
```

Should there perhaps also be a call to ` _cleanup_locked(wkh)`?